### ISSUE TITLE
Allow cancelling programmatic scroll by user

### ISSUE THREAD

**Issue by renegeuze** [cferdinandi/smooth-scroll#309]
It might be something only programmers can feel annoyed about but quite often when a scroll animation is running I'd like to cancel it and scroll the other way around.

A simple but effective method that should work in all modern browsers is to stop the smooth-scroll when the wheel or touchstart event is triggered.

Semi-real code(could be written more compact):

	var cancelEvents  = ['wheel', 'touchstart'];
	var userstopEvent = new Event('smoothscroll_userstop');

	function applyEventFunctionToAll(func) {
		var i;
		for (i=0; i<cancelEvents.length;i++) {
			func('on' + cancelEvents[i]);
		}
	}

	function userStopAnimateScroll() {
		window.dispatchEvent(userstopEvent);
		stopAnimateScroll();
		applyEventFunctionToAll(removeCancelListener);
	}

	function removeCancelListener(eventName) {
		window.removeEventListener(eventName, userStopAnimateScroll);
	}

	function addCancelListener(eventName) {
		window.addEventListener(eventName, userStopAnimateScroll);
	}

	if (allowUserStopAnimationScroll === true) { 
		applyEventFunctionToAll(addCancelListener);
	}

My code does not handle anything like scrollers inside anything other than window. Mostly because I have no clue if it should.

And maybe it does not need to be in there core but it would be nice to be able to add this functionality with the "api".

**Comment by cferdinandi** [comment]
Hey there... to be frank, I'm not sure the utility here is worth the added code bloat. This feels like a feature that doesn't serve much purpose for the average user.

Feel free to fork and modify the code as desired, though.

**Comment by andy-hook** [comment]
Hello, I also wanted to raise a ticket regarding this issue, I think having this will avoid the "jank" that currently exists when you attempt to scroll against the animation, I think this is relevant to many users.

**Comment by cferdinandi** [comment]
Reopening for discussion.

With an approach like this, what happens when someone accidentally clicks or moves the scroll wheel while the animation is in place? Wouldn't this leave them "stranded" on the page somewhere other than where they intended to go?

Or is the idea that if the users clicks/scroll while the animation is in progress, they automatically jump to the anchored location?

**Comment by andy-hook** [comment]
A user interacting with the scroll wheel wants to scroll their window regardless of the current state of the animation. Ideally the animation should end and control returned as soon as the user attempts to scroll themselves.

Basically if the user uses their mouse wheel then the animation should end exactly at that time so that full control of the scroll is resumed.

At the moment you are fighting against the animation if you change your mind and attempt to scroll during the animation.

**Comment by cferdinandi** [comment]
This seems counterintuitive to how a normal anchor link would work. You click the link, you're jumped down to that spot. If you change your mind, well... you're at the anchored location.

**Comment by andy-hook** [comment]
A standard local anchor is an instant interaction though. When animating scroll it is poor user experience to force the user to wait for the entire duration (and it looks really janky)

**Comment by cferdinandi** [comment]
Fair points. In your opinion, what are appropriate events on which to cancel the animation? Scroll events obviously come to mind, but what else? Click? etc.

**Comment by andy-hook** [comment]
mousedown wheel DOMMouseScroll mousewheel keyup touchstart - I wouldn't bind to scroll as that event cannot always be explicitly linked to a user interaction.

**Comment by renegeuze** [comment]
Mousedown might throw off my dad who's one of the billion people that tends to double click buttons because (windows) explorer works like that.
Touch devices never(?) had double tap for opening so it should be less of an issue there.

Although I'm convinced this change is an improvement we can never be sure it suits everybody's needs. So it's good you didn't implement it right after my comment :) Some user testing would also be nice.

**Comment by andy-hook** [comment]
I think it's a really good idea and it's standard practice for me to implement in my projects, yes it shouldn't be bound to click, I highly recommend just the events I mentioned in my last comment.

**Comment by cferdinandi** [comment]
I'd like to do a bit more research on best practices and user expectations before I make any changes here, but I am open to this as a feature.

**Comment by andy-hook** [comment]
Awesome. I forked this code into my current project and added a quick cancel function using '''function cancelScroll() {
      var currentLocation = root.pageYOffset;

      stopAnimateScroll(currentLocation, currentLocation, animationInterval);
    }'''

Not ideal but works for now, would love to see a proper implementation

**Comment by NonesuchStudios** [comment]
Just adding my support for this feature. Scrolling is a very basic, "low-level" user interaction. Any time a user tries to do something that "always works" — like scrolling — and the ui doesn't respond to their input, it's bad UX. I would compare it to preventing a user from switching tabs or closing their browser. It doesn't matter if the user initiated a scroll animation, maybe they did it by accident, or they see content that they want as it scrolls by. 

Thanks for a fantastic script otherwise!

**Comment by cferdinandi** [comment]
I'm still struggling with what user actions should cancel the interaction. I do want to implement this though.

**Comment by NonesuchStudios** [comment]
There's definitely some nuance to that question. I'd say you couldn't go wrong cancelling on scroll wheel and touch events at least. If there's some way to detect interactions with the scrollbar that would also be good (but I'm not sure how you'd go about that.) Keyboard events and mouse clicks get a bit trickier. One option would be to make this configurable? Then if a dev needs to cancel scrolling on some keyup event they'd be able to. Or maybe expose a cancel method that would be bound to arbitrary events?

**Comment by NonesuchStudios** [comment]
You could also bypass this question entirely by checking that the scroll position hasn't changed before each time you run your animation loop. If it has changed, it means that something other than this script impacted the scroll position, implying that the user is trying to scroll the page somehow.

**Comment by NonesuchStudios** [comment]
Something like this maybe:
```
var previousLocation = root.pageYOffset;

var loopAnimateScroll = function () {
  var currentLocation = root.pageYOffset;
  if(previousLocation !== currentLocation) {
    endLocation = currentLocation;
    stopAnimateScroll(position, endLocation, animationInterval);
  }
  timeLapsed += 16;
  percentage = ( timeLapsed / parseInt(animateSettings.speed, 10) );
  percentage = ( percentage > 1 ) ? 1 : percentage;
  position = startLocation + ( distance * easingPattern(animateSettings.easing, percentage) );
  root.scrollTo( 0, Math.floor(position) );
  stopAnimateScroll(position, endLocation, animationInterval);
  previousLocation = root.pageYOffset;
};
```

**Comment by cferdinandi** [comment]
@NonesuchStudios Not necessarily:

> Edit: Just clearing the interval without adjusting focus or firing the callback make sense in this case, since the scroll didn't actually finish.

If they're half-way through a scroll, shouldn't focus be on where they are?

I'm going to consult with a few of my a11y friends on this.

**Comment by NonesuchStudios** [comment]
Interesting point. I would think that since screen readers jump through content in discrete steps it might be less disorienting to stay where you were when the canceled scroll was initiated. I'm far from an expert on these things though, I'd be curious to hear what your friends say!

**Comment by ericwbailey** [comment]
This script is hands-down my favorite scrolling plugin out there, and definitely handles the technological aspects of being accessibility-friendly—my concerns for this issue are more on the cognitive end of things: 

- An interrupt would be disruptive if visual position didn't match programmatic focus. Not all screenreader users are blind, so that discrepancy could get very confusing very fast.
- For less technologically savvy users, a scroll cancel could leave them confused since the intended destination isn't navigated to. I'm especially thinking in terms of motor disabilities or accidental scrolling (say Parkinson's, surprise cat on the trackpad, a janky mousewheel, etc.).

However, a cancel effect would be very handy for tall pages where there's been accidental triggering. Part of me is wondering if there's a threshhold you could set for preventing accidental cancellation, and part of me is thinking that catering to advanced users creates more problems than it solves?

Sorry I don't have a more concrete answer for you.

**Comment by cferdinandi** [comment]
No apologies necessary @ericwbailey.  You hit on most of my major concerns. There's a strong part of me that also feels like as long as the scrolling speed is reasonably fast – say 300 ms or so – then canceling the action hurts more than it helps. 

 Similar to how you can't cancel a normal anchor click - you just have to navigate back up.

**Comment by NonesuchStudios** [comment]
Definitely good points! Given these concerns, my vote would be to add this as an option.

I call animateScroll with a pixel value rather than a selector quite often. When it's used that way there are myriad situations where you'd want to make sure the scroll was interruptible and leave focus untouched.

**Comment by scottaohara** [comment]
So, things that one should consider concerning smooth scroll in general:

This is a feature for sighted users.  As far as my testing of features like this go for keyboard users and especially for users of assistive technology, they (generally) do not care for features like this, as they are typically trying to navigate pages as quickly as possible (using keyboards & menus to navigate a site is far slower than a mouse click, so many of these users have gotten quite adept at trying to move through UIs as quickly as possible).  Smooth scrolling is in opposition to that.  

But regardless of whether this is a feature for them or not, one must keep focus management in mind with scripts like this, as it could be used by all people.

As @ericwbailey said:

> * An interrupt would be disruptive if visual position didn't match programmatic focus. Not all screenreader users are blind, so that discrepancy could get very confusing very fast.   
> * For less technologically savvy users, a scroll cancel could leave them confused since the intended destination isn't navigated to. I'm especially thinking in terms of motor disabilities or accidental scrolling (say Parkinson's, surprise cat on the trackpad, a janky mousewheel, etc.).  

What we have to keep in mind here, is that if there is a cancel of the smooth scroll, what happens to focus management?  Does it still get pushed to the location it should have been?  That seems silly since the action was interrupted.  Should it be reset to the original trigger?  This makes more sense to me...but still seems like a less than ideal solution.  Do we find the first focuable element in the current view port and just attach focus to that?  That...sorta makes sense, but what if there are no focusable elements in the view port?  Or what if there are fixed elements in the view port that would actually move them further up or down the DOM.  Let's thinks some more here...

Again, we should go back to the type of user that would be using smooth scrolling.  

Screen reader users don't use skip links as much as one might think.  Screen readers have built in navigational controls to jump to landmarks, headings and links on a page.  Those controls are how most screen reader users get around a page quickly.  These skip links are mostly just clutter in the list of links that they can jump through with screen reader navigation controls.  

So we probably don't have to worry about screen reader users and them losing focus management, because they're not likely to be using these sorts of skip links.

But as Eric said, not all screen reader users are blind.  Right.  I would classify these people as sighted (or partially sighted) keyboard users.  And again, these are not the target audience for smooth scrolling.  They want to get through the document quickly, and while they might want to use a skip link, it'd arguably be better for them if it just took them to their destination.

So, verbose lead-in out of the way. What should happen here?  

I would submit that the solution to this may be in how the smooth scroll is activated.

If a user activates it via ```keydown```, that means they likely navigated to it via keyboard controls.  In instances like this, I would say the scroll should __not__ be allowed to be canceled as a keyboard user wants to go to the destination of the link.  The smooth scrolling is not the expected behavior for them, as links typically just go to their destination, so they're likely not going to sit there and be like "OH WAIT. THIS DOCUMENT I ALREADY NAVIGATED THROUGH.  LET ME TAB AROUND IN IT AGAIN, RIGHT HERE, BECAUSE LOL".  Also why are they talking in all caps? So mad...  

However, if a user activates the smooth scroll via mouse click/down/whatever, I could very well understand that they might want to cancel out of that functionality.  It may be for reasons like "oh, wait. something shiny caught my eye. I should interact with this brand some more."  Or... no.  It's probably only that one reason.

So, to sum this up.

If a user interacts with the smooth scroll trigger via keyboard controls, I wouldn't suggest you allow the scroll to be cancelled, because it's not expected behavior.

if a user interacts with the smooth scroll trigger via mouse click/down/up/ok, it's probably ok to allow the scroll to be cancelled.  Yes, this could still be a keyboard user with a cat attacking their track pad, or some other slip up.  So to mitigate this, I'd say keep focus on the smooth scroll trigger.  So if the scroll is interrupted, they could just initiate it again.

And there you go @cferdinandi.  Hope you aren't sorry you asked me to join in :)

**Comment by cferdinandi** [comment]
@scottaohara - should keyboard events trigger smooth scrolling at all then?

My love of smooth scrolling: it makes anchor links less disorienting for sighted users by providing a visual reference that you're moving elsewhere on the current page--not going to a new one.

**Comment by cferdinandi** [comment]
Also, skip nags aren't really used that much? I feel like everything I know is a lie.

**Comment by scottaohara** [comment]
skip navs are used.  just not necessarily by blind screen reader users who have different, comparable, means of navigating a page.  this was news to me too, before I started doing more testing with low/no-vision users who never needed them cause they were using the built in screen reader functionality, as i mentioned above.

and yes, though I am quite snarky in my post (sorry it's my default state), I'm not trying to down play it as a UX feature for sighted users.  But let's be clear here, this is a feature enhancement for sighted users, and it would be slowing down the expected interaction for blind keyboard users, if they were to use it.

With that said I don't necessarily think this should be a mouse click only feature, but if it was updated to have an option to make it mouse click only...I'd personally want that.  

But it definitely shouldn't be able to be interrupted if it was initiated by a keyboard user, since the expected functionality is "i click link. I go there", regardless of if this is on the same page or not.  

I'm just saying I'm less worried about mouse interactions cancel out the smooth scroll.  But again, only if a mouse triggered the interaction to begin with.

**Comment by cferdinandi** [comment]
Given that keyboard events also fire as a click, and implementing this would be quite complex for what I perceive to be a limited benefit, I'm labeling this one `wontfix`. Sorry!

**Comment by cferdinandi** [comment]
@NonesuchStudios Revisiting this, can you tell me more about this?

> I call animateScroll with a pixel value rather than a selector quite often. When it's used that way there are myriad situations where you'd want to make sure the scroll was interruptible and leave focus untouched.

**Comment by NonesuchStudios** [comment]
Sure! Some examples from recent projects:

1. To make it clear that a particular page was scrollable, I used animateScroll to scroll down just a little, to provide a "peek" of further content. This is helpful for sighted users, but irrelevant for people navigating with a keyboard or screen reader.
2. To accommodate a large inline video that was exactly the height of the viewport, I used animateScroll to center on the video when a user clicks the play button. Otherwise, it was very difficult to manually scroll perfectly to that position.
3. Scrolling down to content from above the fold when a user clicks a "down arrow" icon. This was a consistent pattern on multiple pages, and for a number reasons it was much easier to scroll to a pixel value, instead of using hash-links. Another feature that's useless for someone who will just tab to the first header below the fold. 

Hopefully those examples make sense!

**Comment by cferdinandi** [comment]
@NonesuchStudios Super helpful, thanks! A few quick follow-up questions, if you don't mind:

1. With the "down arrow" icon, why wouldn't you want to bring focus to the new content? 
2. What should happen after cancelling? Should the page stay where it is? Jump somewhere else?
3. Is this something you feel should happen automatically (ex. when hitting escape or scrolling up) or only run programmatically (via an API call like `smoothScroll.cancelScroll()`)?

**Comment by NonesuchStudios** [comment]
1. In this case, passing focus would be fine actually. The animation should still be user-interruptible though.
2. The page should stay where it is. The whole idea is that if the user tries to initiate their own manual scroll, the script should immediately cede control to the human input instead of "fighting" the user's attempt at scrolling. The only intuitive way to do this is to leave everything as-is. Stop the animation and allow the user to take over, wherever we happen to be.
3. In my ideal world we'd have both: A global option to turn this on and off by default that could be overridden by passing a param to animateScroll. And an api call. The built-in version could be very simple and just cancel the animation when any input causes a scroll. The api would allow the flexibility to handle more complicated setups, any unforeseen edge cases that might crop up.

**Comment by cferdinandi** [comment]
Ok, so the good news: this is coming to version 12 as public method. You'll be able to setup your own events to cancel scrolling using the new `cancelScroll()` method.

The bad news: I will not be adding any automatic support or settings for this. It can only be done via the API, and you'll be responsible for handling things like focus and position on the page.

I'll be releasing Smooth Scroll 12 shortly.

**Comment by NonesuchStudios** [comment]
@cferdinandi fantastic! Thank you so much for your hard work!

**Comment by cferdinandi** [comment]
Done: https://github.com/cferdinandi/smooth-scroll/pull/362
